scylla-cdc/src/checkpoints.rs

```diff
@@ -0,0 +1,352 @@
+use crate::cdc_types::{GenerationTimestamp, StreamID};
+use anyhow;
+use async_trait::async_trait;
+use futures::future::RemoteHandle;
+use futures::FutureExt;
+use scylla::frame::response::result::CqlValue::Timestamp;
+use scylla::prepared_statement::PreparedStatement;
+use scylla::Session;
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::time::sleep;
+
+#[non_exhaustive]
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Checkpoint {
+    pub timestamp: Duration,
+    pub stream_id: StreamID,
+    pub generation: GenerationTimestamp,
+}
+
+/// Spawns a new task that periodically saves checkpoints.
+/// Receives value of the last checkpoint via channel.
+/// We are using [`tokio::sync::watch`] channel, because we only need the latest value.
+/// Returns handle to check on this newly created task.
+pub fn start_saving_checkpoints(
+    checkpoint_saver: Arc<dyn CDCCheckpointSaver>,
+    receiver: tokio::sync::watch::Receiver<Checkpoint>,
+    saving_period: Duration,
+) -> RemoteHandle<anyhow::Result<()>> {
+    let (fut, handle) = async move {
+        loop {
+            let checkpoint = receiver.borrow().clone();
+
+            checkpoint_saver.save_checkpoint(&checkpoint).await?;
+
+            sleep(saving_period).await;
+        }
+    }
+    .remote_handle();
+
+    tokio::task::spawn(fut);
+
+    handle
+}
+
+/// Customizable trait responsible for saving checkpoints.
+#[async_trait]
+pub trait CDCCheckpointSaver: Send + Sync {
+    /// Saves given checkpoint.
+    async fn save_checkpoint(&self, checkpoint: &Checkpoint) -> anyhow::Result<()>;
+    /// When a new generation comes into effect, it saves it into the checkpoint table.
+    async fn save_new_generation(&self, generation: &GenerationTimestamp) -> anyhow::Result<()>;
+    /// Loads last seen generation from the checkpoint table.
+    async fn load_last_generation(&self) -> anyhow::Result<Option<GenerationTimestamp>>;
+    /// Loads last seen timestamp for given stream id from the checkpoint table.
+    async fn load_last_checkpoint(
+        &self,
+        stream_id: &StreamID,
+    ) -> anyhow::Result<Option<chrono::Duration>>;
+}
+
+/// Default implementation for [`CDCCheckpointSaver`] trait.
+pub struct CDCDefaultCheckpointSaver {
+    session: Arc<Session>,
+    checkpoint_table: String,
```

Comment:
When the library switches to the new generation the current set of streams changes, so the new stream readers will save progress in a new set of partitions while the partitions for previous generation are left behind - so the table will grow every time there is a generation change.

We need to add another parameter: TTL. You can write data to scylla with `USING TTL = X` where `X` is the number of seconds after the row will "disappear" and scylla will eventually be able to reclaim the space it uses.

As generation changes aren't supposed to be frequent, you can provide a default TTL of 7 days.